libraries/persistent_store/src/fragment.rs

```diff
@@ -0,0 +1,179 @@
+    range: Range<usize>,
+) -> StoreResult<Option<Vec<u8>>> {
+    let range_len = match range.end.checked_sub(range.start) {
+        None => return Err(StoreError::InvalidArgument),
+        Some(x) => x,
+    };
+    let handles = get_handles(store, keys)?;
+    if handles.is_empty() {
+        return Ok(None);
+    }
+    let mut result = Vec::with_capacity(range_len);
+    let mut offset = 0;
+    for handle in handles {
+        let start = range.start.saturating_sub(offset);
+        let length = handle.get_length(store)?;
+        let end = core::cmp::min(range.end.saturating_sub(offset), length);
+        offset += length;
+        if start < end && end <= length {
+            result.extend(&handle.get_value(store)?[start..end]);
+        }
+    }
+    Ok(Some(result))
+}
+
+/// Writes a value to a sequence of keys as chunks.
+pub fn write(store: &mut Store<impl Storage>, keys: &impl Keys, value: &[u8]) -> StoreResult<()> {
+    let handles = get_handles(store, keys)?;
+    let keys_len = keys.len();
+    let mut updates = Vec::with_capacity(keys_len);
+    let mut chunks = value.chunks(store.max_value_length());
+    for pos in 0..keys_len {
+        let key = keys.key(pos);
+        match (handles.get(pos), chunks.next()) {
+            // No existing handle and no new chunk: nothing to do.
+            (None, None) => (),
+            // Existing handle and no new chunk: remove old handle.
+            (Some(_), None) => updates.push(StoreUpdate::Remove { key }),
+            // Existing handle with same value as new chunk: nothing to do.
+            (Some(handle), Some(value)) if handle.get_value(store)? == value => (),
+            // New chunk: Write (or overwrite) the new value.
+            (_, Some(value)) => updates.push(StoreUpdate::Insert { key, value }),
+        }
+    }
+    if chunks.next().is_some() {
+        // The value is too long.
+        return Err(StoreError::InvalidArgument);
+    }
+    store.transaction(&updates)
+}
+
+/// Deletes the value of a sequence of keys.
+pub fn delete(store: &mut Store<impl Storage>, keys: &impl Keys) -> StoreResult<()> {
+    let updates: Vec<StoreUpdate<Vec<u8>>> = get_handles(store, keys)?
+        .iter()
+        .map(|handle| StoreUpdate::Remove {
+            key: handle.get_key(),
+        })
+        .collect();
+    store.transaction(&updates)
+}
+
+/// Returns the handles of a sequence of keys.
+///
+/// The handles are truncated to the keys that are present.
+fn get_handles(store: &Store<impl Storage>, keys: &impl Keys) -> StoreResult<Vec<StoreHandle>> {
+    let keys_len = keys.len();
+    let mut handles: Vec<Option<StoreHandle>> = vec![None; keys_len as usize];
+    for handle in store.iter()? {
+        let handle = handle?;
+        let pos = match keys.pos(handle.get_key()) {
+            Some(pos) => pos,
+            None => continue,
+        };
+        if pos >= keys_len {
+            return Err(StoreError::InvalidArgument);
+        }
```

Comment:
The problem is that the `Keys` trait is safe (not marked `unsafe`) so anybody can implement it, and they can do mistakes, like returning `Some(out_of_bound_pos)` from `Keys::pos()`. Since I do `handles[pos]` below, I need to check this invariant to avoid panics. The alternative would be to mark the `Keys` trait as `unsafe` such that it's the responsibility of the implementor if this code panics. Since the check is cheap, I prefer to have a safe API.

Reply:
Understood.